src/service/history-impl.js

```diff
@@ -525,6 +552,29 @@ export class HistoryBindingNatural_ {
   }
 
   /**
+   * If this is a hash update the choice of `location.replace` vs
+   * `history.replaceState` is important. Due to bugs, not every browser
+   * triggers `:target` pseudo-class when `replaceState` is called.
+   * See http://www.zachleat.com/web/moving-target/ for more details.
+   * location.replace will trigger a `popstate` event, we temporarily
+   * disable handling it.
+   * @param {string} target
+   *
+   * @override
+   */
+  replaceStateForTarget(target) {
+    this.whenReady_(() => {
+      const hash = target[0] == '#' ? target : `#${target}`;
+      this.ignoreUpcomingPopstate_ = true;
```

Comment:
If the events are synchronous, does it mean that `replace` can fail? Please check. If so, we may need a `try/finally`